pkg/storage/bulk/sst_batcher.go

```diff
@@ -228,22 +264,47 @@ type sstSpan struct {
 // SST spans a split, in which case it is iterated and split into two SSTs, one
 // for each side of the split in the error, and each are retried.
 func AddSSTable(
-	ctx context.Context, db sender, start, end roachpb.Key, sstBytes []byte, disallowShadowing bool,
+	ctx context.Context,
+	db sender,
+	start, end roachpb.Key,
+	sstBytes []byte,
+	disallowShadowing bool,
+	ms *enginepb.MVCCStats,
 ) error {
-	// Create an iterator that iterates over the top level SST for use in stats
-	// computation and to produce split ssts if needed for retries.
+	now := timeutil.Now().UnixNano()
 	iter, err := engine.NewMemSSTIterator(sstBytes, true)
 	if err != nil {
 		return err
 	}
 	defer iter.Close()
 
-	now := timeutil.Now().UnixNano()
-	stats, err := engine.ComputeStatsGo(
-		iter, engine.MVCCKey{Key: start}, engine.MVCCKey{Key: end}, now,
-	)
-	if err != nil {
-		return errors.Wrapf(err, "computing stats for SST [%s, %s)", start, end)
+	var stats enginepb.MVCCStats
+	fast := ms != nil
+	if fast {
+		stats = *ms
+		stats.LastUpdateNanos = now
+	}
+
+	// If we can't use the fast stats path, or race test is enabled,
+	// compute stats over the SST being ingested.
+	if !fast || util.RaceEnabled {
```

Comment:
@dt I remember you mentioning that you would prefer this verification logic in `cmd_add_sstable` but I couldn't convince myself that that fit the current code structure.

With the current design, the stats field we send `cmd_add_sstable` is always non-nil. So there is no logical condition I can think of which lets us differentiate what came from the fast path and what didn't. In this method, however, we can check if the stats passed is non-nil and consider that as having come from the on-the-fly computation. This approach comes with the caveat that someone using this method as a standalone (not with sstbatcher) and testing with race enabled, will incur an unnecessary validation cost.